src/TreeDataGrid.tsx

```diff
@@ -0,0 +1,430 @@
+      const parentRowAndIndex = getParentRowAndIndex(row);
+      if (parentRowAndIndex !== undefined) {
+        event.preventGridDefault();
+        selectCell({ idx, rowIdx: parentRowAndIndex[1] });
+      }
+    }
+
+    if (isCtrlKeyHeldDown(event) && (event.keyCode === 67 || event.keyCode === 86)) {
+      event.preventGridDefault();
+    }
+  }
+
+  function handleRowsChange(updatedRows: R[], { indexes, column }: RowsChangeData<R, SR>) {
+    if (!onRowsChange) return;
+    const updatedRawRows = [...rawRows];
+    const rawIndexes: number[] = [];
+    indexes.forEach((index) => {
+      const rawIndex = rawRows.indexOf(rows[index] as R);
+      updatedRawRows[rawIndex] = updatedRows[index];
+      rawIndexes.push(rawIndex);
+    });
+    onRowsChange(updatedRawRows, {
+      indexes: rawIndexes,
+      column
+    });
+  }
+
+  function toggleGroup(groupId: unknown) {
+    const newExpandedGroupIds = new Set(expandedGroupIds);
+    if (newExpandedGroupIds.has(groupId)) {
+      newExpandedGroupIds.delete(groupId);
+    } else {
+      newExpandedGroupIds.add(groupId);
+    }
+    onExpandedGroupIdsChange(newExpandedGroupIds);
+  }
+
+  function renderRow(
+    key: Key,
+    {
+      row,
+      rowClass,
+      onCellClick,
+      onCellDoubleClick,
+      onCellContextMenu,
+      onRowChange,
+      lastFrozenColumnIndex,
+      copiedCellIdx,
+      draggedOverCellIdx,
+      setDraggedOverRowIdx,
+      selectedCellEditor,
+      selectedCellDragHandle,
+      ...rowProps
+    }: RenderRowProps<R, SR>
+  ) {
+    if (isGroupRow(row)) {
+      const { startRowIndex } = row;
+      return (
+        <GroupedRow
+          key={key}
+          {...rowProps}
+          aria-rowindex={headerAndTopSummaryRowsCount + startRowIndex + 1}
+          row={row}
+          toggleGroup={toggleGroupLatest}
+        />
+      );
+    }
+
+    let ariaRowIndex = rowProps['aria-rowindex'];
+    const parentRowAndIndex = getParentRowAndIndex(row);
+    if (parentRowAndIndex !== undefined) {
+      const { startRowIndex, childRows } = parentRowAndIndex[0];
+      const groupIndex = childRows.indexOf(row);
+      ariaRowIndex = startRowIndex + headerAndTopSummaryRowsCount + groupIndex + 2;
+    }
+
+    return rawRenderRow(key, {
+      ...rowProps,
+      'aria-rowindex': ariaRowIndex,
+      row,
+      rowClass,
+      onCellClick,
+      onCellDoubleClick,
+      onCellContextMenu,
+      onRowChange,
+      lastFrozenColumnIndex,
+      copiedCellIdx,
+      draggedOverCellIdx,
+      setDraggedOverRowIdx,
+      selectedCellEditor,
+      selectedCellDragHandle
+    });
+  }
+
+  return (
+    <DataGrid<R, SR, Key>
+      {...props}
+      role="treegrid"
+      aria-rowcount={
+        rowsCount + 1 + (props.topSummaryRows?.length ?? 0) + (props.bottomSummaryRows?.length ?? 0)
+      }
+      ref={ref}
+      columns={columns}
+      rows={rows as R[]} // TODO: check types
```

Comment:
Can we fix this TODO?
```suggestion
      rows={rows as readonly R[]} // TODO: check types
```

Reply:
I tried but it requires updating column definition. Let's think about grouping props and do it in a separate PR